src/tls.c

```diff
@@ -14924,6 +14924,22 @@ int TLSX_Parse(WOLFSSL* ssl, const byte* input, word16 length, byte msgType,
     else if (!isRequest && ssl->options.haveEMS && !pendingEMS)
         ssl->options.haveEMS = 0;
 #endif
+#if defined(WOLFSSL_TLS13) && !defined(NO_PSK)
+    if (IsAtLeastTLSv1_3(ssl->version) && msgType == server_hello &&
```

Comment:
This belongs in DoTls13ServerHello().
Around tls13.c:5518.

Is there a reason it needs to be in this function?

Reply:
I agree this is for TLS1.3, but that code calls common functions that are in the common TLS code.

For the client, `ssl->extensions` contains TLSX_KEY_SHARE as it has sent it in the ClientHello.
On receipt of the ServerHello which does not contain `TLSX_KEY_SHARE `this option is not found and hence in `TLSX_Parse()`, `KS_PARSE()` (`really TLSX_KeyShare_Parse()`) is not called (tls.c:14825).

As the ServerHello extensions are not explicitly tracked elsewhere outside of `TLSX_Parse()` I put this code in here.

Testing for missing `TLSX_KEY_SHARE `in `DoTls13ServerHello()` at the point you indicate tls13.c:5518 is insufficient as it will always be in `ssl->extensions`.

What is important there is that `ssl->arrays->preMasterSz = 0;` is done.  
I have seen the need to correct the ciphersuite as well, but cannot easily reproduce  that just now.

Reply:
Only need the preMasterSz to be set to 0.
The rest is handled in SetupPskKey().
To better support this, instead of setting preMasterSz here, set ssl->options.noPskDheKe to 1.